### PR TITLE
test: applitools integration for unit tests

### PR DESCRIPTION
**Pull request by vijetmahabaleshwar-okta**
* Adds support to run visual regression tests using [applitools](http://applitools.com/)
* Adds a new script to run test - `yarn test:applitools`
* Adds ability to take screenshots in unit tests using phantomJS browser
* Adds script to upload screenshots to applitools server for comparison with baselines

resolves: OKTA-275723

### CODE REVIEW

File: src/util/UploadImages.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/* global Promise */
+'use strict';
+
+var fs = require('fs');
+var Eyes = require('@applitools/eyes-images').Eyes;
+
+var eyes = new Eyes();
+var ConsoleLogHandler = require('@applitools/eyes-images').ConsoleLogHandler;
+eyes.setLogHandler(new ConsoleLogHandler(true));
+eyes.setApiKey(process.env.APPLITOOLS_API_KEY);
```
Comment (haishengwu-okta) [inline src/util/UploadImages.js:10]: curious how this key be set in travis ?

**Comment by haishengwu-okta** [comment]
also, looks like you're moving a few test file into `visual` folder and only interested a few cases.
- is the goal to enable visual test for all unit tests?
- otherwise, shall we duplicates test just for visual testing? if we go down this path, are we interested to use testcafe (if it's possible for integration)

File: karma.conf.applitools.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+const path = require('path');
+const webpackConfig = require('./webpack.test.config.js');
+const karmaOverrides = require('./test/unit/karma/karma-overrides');
+const rootDir = path.resolve(__dirname);
+
+module.exports = (config) => {
+  const options = {
+    basePath: './',
+    browsers: ['PhantomJSCustom'],
```
[the remark this answers is not in the document]
Comment (vijetmahabaleshwar-okta) [reply karma.conf.applitools.js:12]: I tried searching for solutions to get screenshots in chrome headless. But google recommends using puppeteer for that. I couldn't find anything that could get me on-demand screenshots in chrome. Even applitools example code uses phantom js.

File: src/util/UploadImages.js (hunk 1 above)
Comment (vijetmahabaleshwar-okta) [reply src/util/UploadImages.js:10]: travis secure env vars can be used. It'll be encrypted. We already set okta API tokens in travis for running DevEx SDK tests.

File: karma.conf.applitools.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply karma.conf.applitools.js:12]: I thought phantomjs is dead project.. curious any other suggestions from applitools team?

[the remark this answers is not in the document]
Comment (vijetmahabaleshwar-okta) [reply karma.conf.applitools.js:12]: phantomjs is indeed dead. :(
Applitools recommends using test frameworks like testcafe, protractor, selenium etc to run tests. They have SDKs built on top of these frameworks, which makes it easy to integrate.
Here is a list of frameworks they support - https://applitools.com/docs/api/eyes-sdk/indexm/sdk-platformindex.html

**Comment by vijetmahabaleshwar-okta** [comment]
@haishengwu-okta - Thanks for the comments.
> is the goal to enable visual test for all unit tests?
That's the goal but I've seen some unit tests failing on phantomJS. I'm not sure if we can enable all tests to use this. We can start with a small number of tests and eventually migrate most of them, depending on how stable the tests are.

> otherwise, shall we duplicates test just for visual testing? if we go down this path, are we interested to use testcafe (if it's possible for integration)
I don't want to have any duplicate tests, as it adds to run time, if we want to run both standalone unit tests and visual unit tests on CI. 
Ideally running `yarn test:unit` should run all spec files even those inside `visual` folder. This will not capture screenshots. (And run only on chrome headless)
Running `yarn test:applitools` should run only spec files in `visual` folder. (And run only on phantomJS)

If we can migrate to testcafe easily, we should do that, as applitools has SDK support for that. We won't have to deal with the obsolete phantomJS to get screenshots.

File: karma.conf.applitools.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply karma.conf.applitools.js:12]: Sounds like testcafe would be better choice than phantom. :)

[the remark this answers is not in the document]
Comment (vijetmahabaleshwar-okta) [reply karma.conf.applitools.js:12]: @haishengwu-okta  - But we will need to migrate all the unit tests to use testcafe right. I don't know if that's a trivial effort.

[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply karma.conf.applitools.js:12]: It definitely won't be trivial and would like to chat in widget design meeting.

**Comment by jmelberg-okta** [comment]
Due to the flakiness of adding Applitools to our unit tests, we're going to close this out as a "won't do".

**Pull request closed without merging**